[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/505539367/AI-class/blob/main/Assignment%203%20-%20Ying%20Yang.ipynb)

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## **Part 1. Loading the dataset**

In [2]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
# Output the first 15 rows of the data
dataset.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
dataset.info()
dataset.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

In [4]:
dataset.size #number of datapoint

3000

## **Part 2. Splitting the dataset**

In [5]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[["Temperature °C", "Mols KCL"]]
X #Features

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [6]:
y = dataset[["Size nm^3"]]
y #labels

Size nm^3
0    6.244743e+05
1    5.779610e+05
2    6.196847e+05
3    1.460449e+06
4    4.325726e+04
..            ...
995  1.545661e+06
996  6.737041e+05
997  3.477543e+05
998  8.684794e+05
999  8.476413e+05

[1000 rows x 1 columns]

In [7]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1)
X_train

Temperature °C  Mols KCL
848             772       382
607              67       331
343             520        76
374             353       602
245             478       511
..              ...       ...
708              48       223
834             361         6
861             704        61
823             410       511
522             247       953

[900 rows x 2 columns]

In [8]:
X_test

Temperature °C  Mols KCL
992             436       543
45               75       379
350             552       967
228             836       573
880             900       281
..              ...       ...
87              846        30
218             104       813
107             675       759
669             427       564
923             381       493

[100 rows x 2 columns]

In [9]:
y_train

Size nm^3
848  603241.257100
607   48288.314290
343   85445.028570
374  439602.400000
245  501712.600000
..             ...
708   23404.828570
834    8665.028571
861   94442.314290
823  431400.600000
522  499694.828600

[900 rows x 1 columns]

In [10]:
y_test

Size nm^3
992  4.871523e+05
45   6.185403e+04
350  1.100909e+06
228  9.774688e+05
880  5.188560e+05
..            ...
87   6.093771e+04
218  1.892368e+05
107  1.049209e+06
669  4.958685e+05
923  3.871823e+05

[100 rows x 1 columns]

## **Part 3. Perform a Linear Regression**

In [11]:
# Use sklearn to train a model on the training set

regr_model = LinearRegression()# Create linear regression object

regr_model.fit(X_train, y_train)# Train the model using the training sets

# Create a sample datapoint and predict the output of that sample with the trained model

regr_model.predict(X_test) # Make predictions


array([[ 523807.98545639],
       [  36703.97594083],
       [1059916.83783607],
       [ 908432.24848154],
       [ 666548.75512808],
       [ 732393.15724551],
       [ 327298.87311439],
       [ 999679.4284511 ],
       [ 659330.11448511],
       [ 540765.16571406],
       [ 257983.02418335],
       [1038122.93579159],
       [ -98697.52709448],
       [ 414876.8642796 ],
       [ 622319.09412031],
       [ 384155.19857308],
       [ 557989.41592637],
       [-233452.39016661],
       [ 418241.48311136],
       [   7585.83220528],
       [ 775229.60935085],
       [1059970.251827  ],
       [  11943.05500749],
       [  63116.16076404],
       [ 490175.38193491],
       [ 576304.97215146],
       [ 105862.58788815],
       [ 466243.81801017],
       [ 103646.03195525],
       [ 942739.08798931],
       [ 566787.8765339 ],
       [ 135820.33053324],
       [1005518.65820321],
       [1290067.56866674],
       [ 461398.41916227],
       [ 211050.40925231],
       [ 556802.07495424],
 

In [12]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = regr_model.score(X_train, y_train)
print("Score for model :", score)
print("Standard devisation :", score.std())

Score for model : 0.8589677411427145
Standard devisation : 0.0


### Score:
The score of the model is between 0 to 1 (0% to 100%). So if it is 100%, the two variables are perfectly correlated. If it is 0%, the two variables are not correlated at all. The score for the model above is around 0.86, two variable has 86% correlation.

In [13]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("Coeffients :", regr_model.coef_) 
print("Intercept :", regr_model.intercept_) 

Coeffients : [[ 884.88696094 1022.31595498]]
Intercept : [-417120.29306632]


### Equation for h(x):
$h(x) = 884.88696094x_0 + 1022.31595498x_1 - 417120.29306632$

where $x_0$ is Temperature °C and $x_1$ is Mols KCL.

## **Part 4. Use Cross Validation**

In [14]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# Report on their finding and their significance

print("Use the cross_val_score with 3 folds:")

scores = cross_val_score(regr_model ,X_train, y_train, cv = 3)
print("\tCross_val_scores :",scores)
print("\tAccracy :",scores.mean())
print("\tStandard devisation :", scores.std())

Use the cross_val_score with 3 folds:
	Cross_val_scores : [0.86352937 0.86278877 0.84763989]
	Accracy : 0.8579860112757928
	Standard devisation : 0.00732205654777474


In [15]:
print("Use the cross_val_score with 5 folds:")

scores = cross_val_score(regr_model ,X_train, y_train, cv = 5)
print("\tCross_val_scores :",scores)
print("\tAccracy :",scores.mean())
print("\tStandard devisation :", scores.std())

Use the cross_val_score with 5 folds:
	Cross_val_scores : [0.87193601 0.84785008 0.86314274 0.85432522 0.85000918]
	Accracy : 0.8574526477254254
	Standard devisation : 0.008939127809679261


In [16]:
print("Use the cross_val_score with 7 folds:")

scores = cross_val_score(regr_model ,X_train, y_train, cv = 7)
print("\tCross_val_scores :",scores)
print("\tAccracy :",scores.mean())
print("\tStandard devisation :", scores.std())

Use the cross_val_score with 7 folds:
	Cross_val_scores : [0.87055772 0.85354048 0.86289012 0.86337092 0.83123724 0.84123179
 0.87062513]
	Accracy : 0.8562076281825203
	Standard devisation : 0.013962539659427873


### Cross Validation finding:
Two variable always has around 86% correlation, and standard devisation is increasing as I tried to split the data into more partes.

## **Part 5. Using Polynomial Regression**

In [17]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
feature = poly.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(feature, y, test_size = 0.1) # split the feature and label
poly_model = regr_model
poly_model.fit(X_train, y_train)

# Make predictions
poly_model.predict(X_test) 

array([[ 311881.25713919],
       [  35586.25713649],
       [  82132.71429121],
       [ 148395.45713194],
       [  49376.45714586],
       [ 942241.45714799],
       [1113796.45715487],
       [1250746.40001711],
       [ 630814.02857756],
       [ 803035.85714748],
       [ 278615.42856792],
       [1143874.3142887 ],
       [   6655.11429234],
       [  57363.42857545],
       [ 288919.42856226],
       [  39528.82856425],
       [ 153021.45713501],
       [1707616.6000362 ],
       [ 454249.25713397],
       [  63602.40000423],
       [1277493.8286057 ],
       [1743673.25718368],
       [ 412822.25713374],
       [  13603.11428718],
       [ 513347.11429222],
       [ 723400.31428113],
       [1086512.31429547],
       [ 880350.42856408],
       [1081124.11428292],
       [ 555003.42856295],
       [   5175.40001588],
       [ 422777.82856259],
       [ 502017.0285632 ],
       [ 920531.31429656],
       [ 360092.31428011],
       [ 458609.45713477],
       [ 398301.2571338 ],
 

In [18]:
# Report on the metrics and output the resultant equation as you did in Part 3.
score = poly_model.score(X_train, y_train)
print("Score for model :", score)
print("Coeffients :", poly_model.coef_) 
print("Intercept :", poly_model.intercept_) 

Score for model : 1.0
Coeffients : [[ 0.00000000e+00  1.20000000e+01 -1.28053937e-07 -2.20944827e-12
   2.00000000e+00  2.85714287e-02]]
Intercept : [1.90200517e-05]


### Equation for h(x):

Two coeffients : $-1.28053937e-07$  and $-2.20944827e-12$ are very small(close to 0) so will ignored it.

$h(x) = (1.2 \times 10^1)x_0 + (2.0)x_0x_1 + (2.85714287\times 10^{-2})x^{2}_1 - 1.90200517\times 10^{-5}$

where $x_0$ is Temperature °C and $x_1$ is Mols KCL.

### Score:
The score of model using polynomail regression is 1.0, means two variable 100% correlated.